File: backend/handlers/jadwal_pertemuan.go
package handlers

import (
	"net/http"
	"strconv"
	"time"

	"stia-bayuangga/backend/config"
	"stia-bayuangga/backend/models"

	"github.com/gin-gonic/gin"
)

// ===================== JADWAL PERTEMUAN (Schedule per meeting) =====================

// GetJadwalPertemuan returns the 14-meeting schedule for a specific course+kelas
// GET /api/jadwal-pertemuan/:mata_kuliah_id?kelas=A
func GetJadwalPertemuan(c *gin.Context) {
	mkIDStr := c.Param("mata_kuliah_id")
	mkID, _ := strconv.ParseUint(mkIDStr, 10, 32)
	kelas := c.DefaultQuery("kelas", "A")

	var items []models.JadwalPertemuan
	config.DB.Where("mata_kuliah_id = ? AND kelas = ?", mkID, kelas).
		Order("pertemuan asc").Find(&items)

	// Get MK info
	var mk models.MataKuliah
	config.DB.First(&mk, mkID)

	c.JSON(http.StatusOK, gin.H{
		"mata_kuliah_id": mkID,
		"mata_kuliah":    mk,
		"kelas":          kelas,
		"count":          len(items),
		"data":           items,
	})
}

// GetAllJadwalPertemuan returns schedules for all courses (for student view)
// GET /api/jadwal-pertemuan/all?nim=2024101001
func GetAllJadwalPertemuan(c *gin.Context) {
	nim := c.Query("nim")

	// If NIM provided, get only the MK this student is enrolled in
	if nim != "" {
		var krsItems []models.KRS
		config.DB.Where("nim = ?", nim).Find(&krsItems)

		mkIDs := []uint{}
		for _, k := range krsItems {
			mkIDs = append(mkIDs, k.MataKuliahID)
		}

		var items []models.JadwalPertemuan
		if len(mkIDs) > 0 {
			config.DB.Preload("MataKuliah").Where("mata_kuliah_id IN ?", mkIDs).
				Order("mata_kuliah_id asc, pertemuan asc").Find(&items)
		}

		c.JSON(http.StatusOK, gin.H{
			"nim":   nim,
			"count": len(items),
			"data":  items,
		})
		return
	}

	// Otherwise return all — preload MataKuliah for kode lookup
	var items []models.JadwalPertemuan
	config.DB.Preload("MataKuliah").Order("mata_kuliah_id asc, pertemuan asc").Find(&items)

	c.JSON(http.StatusOK, gin.H{
		"count": len(items),
		"data":  items,
	})
}

// SaveJadwalPertemuan saves/updates a single meeting schedule
// POST /api/jadwal-pertemuan
// Body: { "kode_mk": "ANA 146", "kelas": "B", "pertemuan": 3, "tanggal": "2026-03-18",
//         "tanggal_asli": "2026-03-16", "jam_mulai": "13:00", "jam_selesai": "14:30",
//         "mode": "online", "catatan": "Kelas Pengganti - Dosen Izin" }
func SaveJadwalPertemuan(c *gin.Context) {
	var input struct {
		MataKuliahID uint   `json:"mata_kuliah_id"`
		KodeMK       string `json:"kode_mk"`
		Kelas        string `json:"kelas"`
		Pertemuan    int    `json:"pertemuan" binding:"required"`
		Tanggal      string `json:"tanggal"`
		TanggalAsli  string `json:"tanggal_asli"`
		JamMulai     string `json:"jam_mulai"`
		JamSelesai   string `json:"jam_selesai"`
		Mode         string `json:"mode"`
		Catatan      string `json:"catatan"`
		UpdatedBy    string `json:"updated_by"`
	}
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Resolve MataKuliahID from kode_mk if not provided
	if input.MataKuliahID == 0 && input.KodeMK != "" {
		var mk models.MataKuliah
		if err := config.DB.Where("kode = ?", input.KodeMK).First(&mk).Error; err == nil {
			input.MataKuliahID = mk.ID
		} else {
			// Auto-create MataKuliah if not found
			mk = models.MataKuliah{Kode: input.KodeMK, Nama: input.KodeMK, SKS: 3}
			config.DB.Create(&mk)
			input.MataKuliahID = mk.ID
		}
	}
	if input.MataKuliahID == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "mata_kuliah_id or kode_mk is required"})
		return
	}

	if input.Kelas == "" {
		input.Kelas = "A"
	}
	if input.Mode == "" {
		input.Mode = "offline"
	}

	// Upsert: find by mata_kuliah_id + kelas + pertemuan
	var jp models.JadwalPertemuan
	result := config.DB.Where("mata_kuliah_id = ? AND kelas = ? AND pertemuan = ?",
		input.MataKuliahID, input.Kelas, input.Pertemuan).First(&jp)

	if result.Error != nil {
		// Create new
		jp = models.JadwalPertemuan{
			MataKuliahID: input.MataKuliahID,
			KodeMK:       input.KodeMK,
			Kelas:        input.Kelas,
			Pertemuan:    input.Pertemuan,
			Tanggal:      input.Tanggal,
			TanggalAsli:  input.TanggalAsli,
			JamMulai:     input.JamMulai,
			JamSelesai:   input.JamSelesai,
			Mode:         input.Mode,
			Catatan:      input.Catatan,
			UpdatedBy:    input.UpdatedBy,
			UpdatedAt:    time.Now(),
		}
		config.DB.Create(&jp)
	} else {
		// Update existing
		config.DB.Model(&jp).Updates(map[string]interface{}{
			"tanggal":      input.Tanggal,
			"tanggal_asli": input.TanggalAsli,
			"jam_mulai":    input.JamMulai,
			"jam_selesai":  input.JamSelesai,
			"mode":         input.Mode,
			"catatan":      input.Catatan,
			"updated_by":   input.UpdatedBy,
			"updated_at":   time.Now(),
		})
	}

	c.JSON(http.StatusOK, gin.H{"message": "Jadwal pertemuan disimpan", "data": jp})
}

// BulkSaveJadwalPertemuan saves all 14 meetings at once
// POST /api/jadwal-pertemuan/bulk
// Body: { "kode_mk": "ANA 146", "kelas": "B", "updated_by": "BAP",
//         "items": [{ "pertemuan": 1, "tanggal": "2026-03-02", ... }, ...] }
func BulkSaveJadwalPertemuan(c *gin.Context) {
	var input struct {
		MataKuliahID uint   `json:"mata_kuliah_id"`
		KodeMK       string `json:"kode_mk"`
		Kelas        string `json:"kelas"`
		UpdatedBy    string `json:"updated_by"`
		Items        []struct {
			Pertemuan   int    `json:"pertemuan"`
			Tanggal     string `json:"tanggal"`
			TanggalAsli string `json:"tanggal_asli"`
			JamMulai    string `json:"jam_mulai"`
			JamSelesai  string `json:"jam_selesai"`
			Mode        string `json:"mode"`
			Catatan     string `json:"catatan"`
		} `json:"items"`
	}
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Resolve MataKuliahID from kode_mk if not provided
	if input.MataKuliahID == 0 && input.KodeMK != "" {
		var mk models.MataKuliah
		if err := config.DB.Where("kode = ?", input.KodeMK).First(&mk).Error; err == nil {
			input.MataKuliahID = mk.ID
		} else {
			// Auto-create MataKuliah if not found
			mk = models.MataKuliah{Kode: input.KodeMK, Nama: input.KodeMK, SKS: 3}
			config.DB.Create(&mk)
			input.MataKuliahID = mk.ID
		}
	}
	if input.MataKuliahID == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "mata_kuliah_id or kode_mk is required"})
		return
	}

	if input.Kelas == "" {
		input.Kelas = "A"
	}

	saved := 0
	for _, item := range input.Items {
		if item.Pertemuan < 1 || item.Pertemuan > 14 {
			continue
		}
		mode := item.Mode
		if mode == "" {
			mode = "offline"
		}

		var jp models.JadwalPertemuan
		result := config.DB.Where("mata_kuliah_id = ? AND kelas = ? AND pertemuan = ?",
			input.MataKuliahID, input.Kelas, item.Pertemuan).First(&jp)

		if result.Error != nil {
			jp = models.JadwalPertemuan{
				MataKuliahID: input.MataKuliahID,
				KodeMK:       input.KodeMK,
				Kelas:        input.Kelas,
				Pertemuan:    item.Pertemuan,
				Tanggal:      item.Tanggal,
				TanggalAsli:  item.TanggalAsli,
				JamMulai:     item.JamMulai,
				JamSelesai:   item.JamSelesai,
				Mode:         mode,
				Catatan:      item.Catatan,
				UpdatedBy:    input.UpdatedBy,
				UpdatedAt:    time.Now(),
			}
			config.DB.Create(&jp)
		} else {
			config.DB.Model(&jp).Updates(map[string]interface{}{
				"tanggal":      item.Tanggal,
				"tanggal_asli": item.TanggalAsli,
				"jam_mulai":    item.JamMulai,
				"jam_selesai":  item.JamSelesai,
				"mode":         mode,
				"catatan":      item.Catatan,
				"updated_by":   input.UpdatedBy,
				"updated_at":   time.Now(),
			})
		}
		saved++
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "Jadwal pertemuan bulk disimpan",
		"saved":   saved,
	})
}

File: api/jadwal-pertemuan.php
<?php
// ============================================
// Jadwal Pertemuan API Handlers
// Port of Go handlers to PHP + MySQL
// ============================================

/**
 * GET /api/jadwal-pertemuan/all
 */
function getAllJadwalPertemuan() {
    $db = getDB();
    $nim = $_GET['nim'] ?? '';

    if ($nim) {
        // Get MK IDs from KRS for this student
        $stmt = $db->prepare('SELECT mata_kuliah_id FROM krs WHERE nim = ?');
        $stmt->execute([$nim]);
        $mkIds = $stmt->fetchAll(PDO::FETCH_COLUMN);

        if (empty($mkIds)) {
            jsonResponse(['nim' => $nim, 'count' => 0, 'data' => []]);
        }

        $placeholders = implode(',', array_fill(0, count($mkIds), '?'));
        $stmt = $db->prepare("
            SELECT jp.*, mk.kode as mk_kode, mk.nama as mk_nama, mk.sks as mk_sks
            FROM jadwal_pertemuan jp
            LEFT JOIN mata_kuliah mk ON jp.mata_kuliah_id = mk.id
            WHERE jp.mata_kuliah_id IN ($placeholders)
            ORDER BY jp.mata_kuliah_id ASC, jp.pertemuan ASC
        ");
        $stmt->execute($mkIds);
        $items = $stmt->fetchAll();

        jsonResponse(['nim' => $nim, 'count' => count($items), 'data' => formatItems($items)]);
    }

    // Return all
    $stmt = $db->query("
        SELECT jp.*, mk.kode as mk_kode, mk.nama as mk_nama, mk.sks as mk_sks
        FROM jadwal_pertemuan jp
        LEFT JOIN mata_kuliah mk ON jp.mata_kuliah_id = mk.id
        ORDER BY jp.mata_kuliah_id ASC, jp.pertemuan ASC
    ");
    $items = $stmt->fetchAll();

    jsonResponse(['count' => count($items), 'data' => formatItems($items)]);
}

/**
 * GET /api/jadwal-pertemuan/:id
 */
function getJadwalPertemuan($mkId) {
    $db = getDB();
    $kelas = $_GET['kelas'] ?? 'A';

    $stmt = $db->prepare("
        SELECT jp.*, mk.kode as mk_kode, mk.nama as mk_nama, mk.sks as mk_sks
        FROM jadwal_pertemuan jp
        LEFT JOIN mata_kuliah mk ON jp.mata_kuliah_id = mk.id
        WHERE jp.mata_kuliah_id = ? AND jp.kelas = ?
        ORDER BY jp.pertemuan ASC
    ");
    $stmt->execute([$mkId, $kelas]);
    $items = $stmt->fetchAll();

    // Get MK info
    $mkStmt = $db->prepare('SELECT * FROM mata_kuliah WHERE id = ?');
    $mkStmt->execute([$mkId]);
    $mk = $mkStmt->fetch() ?: null;

    jsonResponse([
        'mata_kuliah_id' => $mkId,
        'mata_kuliah'    => $mk,
        'kelas'          => $kelas,
        'count'          => count($items),
        'data'           => formatItems($items),
    ]);
}

/**
 * POST /api/jadwal-pertemuan
 */
function saveJadwalPertemuan() {
    $db = getDB();
    $input = getJsonBody();

    $mkId = $input['mata_kuliah_id'] ?? 0;
    $kodeMK = $input['kode_mk'] ?? '';

    // Resolve MataKuliahID from kode_mk if needed
    if (!$mkId && $kodeMK) {
        $stmt = $db->prepare('SELECT id FROM mata_kuliah WHERE kode = ?');
        $stmt->execute([$kodeMK]);
        $row = $stmt->fetch();
        if ($row) {
            $mkId = $row['id'];
        } else {
            // Auto-create
            $stmt = $db->prepare('INSERT INTO mata_kuliah (kode, nama, sks) VALUES (?, ?, 3)');
            $stmt->execute([$kodeMK, $kodeMK]);
            $mkId = $db->lastInsertId();
        }
    }

    if (!$mkId) {
        jsonResponse(['error' => 'mata_kuliah_id or kode_mk is required'], 400);
    }

    $kelas = $input['kelas'] ?: 'A';
    $pertemuan = (int)($input['pertemuan'] ?? 0);
    $mode = $input['mode'] ?: 'offline';

    if (!$pertemuan) {
        jsonResponse(['error' => 'pertemuan is required'], 400);
    }

    // Upsert
    $stmt = $db->prepare('SELECT id FROM jadwal_pertemuan WHERE mata_kuliah_id = ? AND kelas = ? AND pertemuan = ?');
    $stmt->execute([$mkId, $kelas, $pertemuan]);
    $existing = $stmt->fetch();

    if ($existing) {
        $stmt = $db->prepare("
            UPDATE jadwal_pertemuan SET
                tanggal = ?, tanggal_asli = ?, jam_mulai = ?, jam_selesai = ?,
                mode = ?, catatan = ?, updated_by = ?, updated_at = NOW()
            WHERE id = ?
        ");
        $stmt->execute([
            $input['tanggal'] ?? '', $input['tanggal_asli'] ?? '',
            $input['jam_mulai'] ?? '', $input['jam_selesai'] ?? '',
            $mode, $input['catatan'] ?? '', $input['updated_by'] ?? '',
            $existing['id']
        ]);
        $id = $existing['id'];
    } else {
        $stmt = $db->prepare("
            INSERT INTO jadwal_pertemuan
                (mata_kuliah_id, kode_mk, kelas, pertemuan, tanggal, tanggal_asli, jam_mulai, jam_selesai, mode, catatan, updated_by)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ");
        $stmt->execute([
            $mkId, $kodeMK, $kelas, $pertemuan,
            $input['tanggal'] ?? '', $input['tanggal_asli'] ?? '',
            $input['jam_mulai'] ?? '', $input['jam_selesai'] ?? '',
            $mode, $input['catatan'] ?? '', $input['updated_by'] ?? ''
        ]);
        $id = $db->lastInsertId();
    }

    // Return saved record
    $stmt = $db->prepare('SELECT * FROM jadwal_pertemuan WHERE id = ?');
    $stmt->execute([$id]);
    $jp = $stmt->fetch();

    jsonResponse(['message' => 'Jadwal pertemuan disimpan', 'data' => $jp]);
}

/**
 * POST /api/jadwal-pertemuan/bulk
 */
function bulkSaveJadwalPertemuan() {
    $db = getDB();
    $input = getJsonBody();

    $mkId = $input['mata_kuliah_id'] ?? 0;
    $kodeMK = $input['kode_mk'] ?? '';
    $kelas = $input['kelas'] ?: 'A';
    $updatedBy = $input['updated_by'] ?? 'BAP';
    $items = $input['items'] ?? [];

    if (empty($items)) {
        jsonResponse(['error' => 'items is required'], 400);
    }

    // Resolve MataKuliahID
    if (!$mkId && $kodeMK) {
        $stmt = $db->prepare('SELECT id FROM mata_kuliah WHERE kode = ?');
        $stmt->execute([$kodeMK]);
        $row = $stmt->fetch();
        if ($row) {
            $mkId = $row['id'];
        } else {
            $stmt = $db->prepare('INSERT INTO mata_kuliah (kode, nama, sks) VALUES (?, ?, 3)');
            $stmt->execute([$kodeMK, $kodeMK]);
            $mkId = $db->lastInsertId();
        }
    }

    if (!$mkId) {
        jsonResponse(['error' => 'mata_kuliah_id or kode_mk is required'], 400);
    }

    $saved = 0;
    $db->beginTransaction();

    try {
        foreach ($items as $item) {
            $pertemuan = (int)($item['pertemuan'] ?? 0);
            if (!$pertemuan) continue;

            $stmt = $db->prepare('SELECT id FROM jadwal_pertemuan WHERE mata_kuliah_id = ? AND kelas = ? AND pertemuan = ?');
            $stmt->execute([$mkId, $kelas, $pertemuan]);
            $existing = $stmt->fetch();

            if ($existing) {
                $stmt = $db->prepare("
                    UPDATE jadwal_pertemuan SET
                        kode_mk = ?, tanggal = ?, tanggal_asli = ?, jam_mulai = ?, jam_selesai = ?,
                        mode = ?, catatan = ?, updated_by = ?, updated_at = NOW()
                    WHERE id = ?
                ");
                $stmt->execute([
                    $kodeMK, $item['tanggal'] ?? '', $item['tanggal_asli'] ?? '',
                    $item['jam_mulai'] ?? '', $item['jam_selesai'] ?? '',
                    $item['mode'] ?? 'offline', $item['catatan'] ?? '', $updatedBy,
                    $existing['id']
                ]);
            } else {
                $stmt = $db->prepare("
                    INSERT INTO jadwal_pertemuan
                        (mata_kuliah_id, kode_mk, kelas, pertemuan, tanggal, tanggal_asli, jam_mulai, jam_selesai, mode, catatan, updated_by)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ");
                $stmt->execute([
                    $mkId, $kodeMK, $kelas, $pertemuan,
                    $item['tanggal'] ?? '', $item['tanggal_asli'] ?? '',
                    $item['jam_mulai'] ?? '', $item['jam_selesai'] ?? '',
                    $item['mode'] ?? 'offline', $item['catatan'] ?? '', $updatedBy
                ]);
            }
            $saved++;
        }

        $db->commit();
    } catch (Exception $e) {
        $db->rollBack();
        jsonResponse(['error' => 'Bulk save failed: ' . $e->getMessage()], 500);
    }

    // Return all saved items
    $stmt = $db->prepare("
        SELECT jp.*, mk.kode as mk_kode, mk.nama as mk_nama, mk.sks as mk_sks
        FROM jadwal_pertemuan jp
        LEFT JOIN mata_kuliah mk ON jp.mata_kuliah_id = mk.id
        WHERE jp.mata_kuliah_id = ? AND jp.kelas = ?
        ORDER BY jp.pertemuan ASC
    ");
    $stmt->execute([$mkId, $kelas]);
    $allItems = $stmt->fetchAll();

    jsonResponse([
        'message' => "$saved jadwal pertemuan disimpan",
        'mata_kuliah_id' => $mkId,
        'kode_mk' => $kodeMK,
        'kelas' => $kelas,
        'count' => count($allItems),
        'data' => formatItems($allItems),
    ]);
}

/**
 * Format items to match Go API response structure
 */
function formatItems($items) {
    return array_map(function($row) {
        $result = [
            'id' => (int)$row['id'],
            'mata_kuliah_id' => (int)$row['mata_kuliah_id'],
            'kode_mk' => $row['kode_mk'],
            'kelas' => $row['kelas'],
            'pertemuan' => (int)$row['pertemuan'],
            'tanggal' => $row['tanggal'],
            'tanggal_asli' => $row['tanggal_asli'],
            'jam_mulai' => $row['jam_mulai'],
            'jam_selesai' => $row['jam_selesai'],
            'mode' => $row['mode'],
            'catatan' => $row['catatan'],
            'updated_by' => $row['updated_by'],
            'updated_at' => $row['updated_at'],
        ];
        // Include mata_kuliah info if available
        if (isset($row['mk_kode'])) {
            $result['mata_kuliah'] = [
                'id' => (int)$row['mata_kuliah_id'],
                'kode' => $row['mk_kode'],
                'nama' => $row['mk_nama'],
                'sks' => (int)($row['mk_sks'] ?? 3),
            ];
        }
        return $result;
    }, $items);
}
